### DISCUSSION TITLE
Terrastation TS-XL

### DISCUSSION THREAD

**Discussion by jhmc93** [1000001101000/Debian_on_Buffalo#181] [General]
Hello I ended up buying the terrastation ts-xl, 
I’m just wondering if theres a guide for this specific device?
Regards

**Comment by 1000001101000** [comment]
There’s no extra steps for this device beyond what’s in the instructions for setting up /boot etc. It is one of the models that will only work with the custom kernel image that is part of this project but the installer sets that up automatically.

**Reply by jhmc93** [reply]
Will it be ok to run the Debian bullseye installer, or do I need to install buster first?

**Comment by 1000001101000** [comment]
Yes! There is no reason that I’m aware of to run Buster at this point.

In fact, I’ll probably end up removing Buster support when Bookworm becomes the new “stable”. I haven’t decided yet but maintaining the installers for older systems get to be annoying with little/no benefit

**Reply by jhmc93** [reply]
So I can install bullseye?.
I don't need to install buster then update to bullseye, 
I'm only asking as it says this:
replace the armel bullseye files with the ones that last worked properly with Orion5x devices

**Comment by 1000001101000** [comment]
yes, that is the commit message from me fixing an issue with the bullseye installer, as far as I know that issue remains fixed.

**Reply by jhmc93** [reply]
I brought the TS-XL off ebay but it had no HDD with it and I'm getting the red emergency mode error due to putting in a clean HDD, is it possible to run the debian installer still?

**Comment by 1000001101000** [comment]
The wiki instructions include how to minimally partition and format a drive so that you can load the installer files onto it and move forward with an install. If you would prefer to follow the method that starts with the device running the stock firmware you could follow the procedure to install the stock firmware, then install debian.

Either way you have options for moving forward. 

Also, if you would like to skip straight to Debian Bookworm rather than Debian Bullseye I generated my first set of installer images for Bookworm yesterday and was able to install Bookworm on my TS-XL without any issues. I should be able to upload them to this issue if you want to give them a try.

**Reply by jhmc93** [reply]
do they need to setup with raid?

**Reply by 1000001101000** [reply]
The wiki information covers what's required to boot the device. nothing has to be RAID if you don't want it to be, if you do want that it has to meet certain criteria.

basically /boot must be:

- on partition `#1` (or each partition1 that make up the raid)
- a regular partition or an mdadm raid1 with metadata version 0.9
- formatted ext3, for this model device and older the inode size must be 128
- sufficiently small for the uboot your device runs, 1GB or less should be fine for that model.

**Reply by jhmc93** [reply]
do i just copy the boot files for this device over? Have u had any problems with bookworm? as i'd like to be on the latest version of Debian?, I was having trouble mounting /boot to my HDD but i've done it now

**Reply by 1000001101000** [reply]
I haven't had any problems with Bookworm on the TS-XL so far. I've been running Bookworm on my laptop and x86 NAS's for a few months without any memorable issues.

**Reply by jhmc93** [reply]
I'm having problems with the /boot mount everytime i restart my nas it unmounts itself, any ideas?

**Comment by 1000001101000** [comment]
actually, it looks like the release date for bookworm has been set to 6/10. I'll probably have the first round of installer images on github around the same time.

**Comment by 1000001101000** [comment]
[tsxl-bookworm.zip](https://github.com/1000001101000/Debian_on_Buffalo/files/11669434/tsxl-bookworm.zip)
These are the files I used for my Bookworm install on my TS-XL

**Reply by jhmc93** [reply]
so it pretty much the same replacing the boot files with the ones in that zip?

**Reply by jhmc93** [reply]
ye my hard drive fails to mount on startup any ideas?

**Reply by 1000001101000** [reply]
start up of what? fails to mount where?

**Reply by jhmc93** [reply]
When I start my drive it fails to automatically mount my /boot partition on start up, I got Debian bookworm installer up little things though,

> Where does it store data for example if I install webmin, is it in the /boot partition or the / partition

> Do I need a swap partition?

Thanks

**Reply by 1000001101000** [reply]
I still don’t understand the comment about mounting /boot, but if you’re currently in the installer it likely doesn’t matter.

when you get to the partitioning section you will specify what partition/etc will be configured as /boot as well as which will be / and any other filesystems you want to configure. If you would like, you can post a screenshot of your partition screen before you move on to the next step so I can validate it. failing to properly set up the partitions can lead to the debian install failing to boot successfully.

you will almost certainly want to have swap set up for this device.

I haven’t used webmin much, I’ve heard it doesn’t run well on these ARM devices. It can also make changes on your behalf that break things in ways that can be hard to troubleshoot/recover from.  

admittedly, I mostly only hear from people who run into problems so my information is a bit skewed.

**Reply by jhmc93** [reply]
nvm, i'll send u a screenshot of my partition page once i have the installer running again, I'm using webmin as an example as I was wondering if u install it what partition does it store it on, For example if u set up a VM on bare metal PC and u give it a HDD space of 32gb to store data, I just wondering what partition stores that data e.g apt packages and other programs

**Reply by 1000001101000** [reply]
the answer to that question depends on how you set up your partitions.

generally everything will go in the “/“ partition unless you’ve specified otherwise.

**Reply by jhmc93** [reply]
Hey,
Here's my partition table, bare in mind I'm starting from a Blank HDD 
![Screenshot_2](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/20615da2-9aa0-4a39-aeeb-c80b94944076)

**Reply by 1000001101000** [reply]
I believe that should work

**Reply by jhmc93** [reply]
does the lcd display  work like  it should on the front of the nas when the install is done?

**Reply by 1000001101000** [reply]
All I've implemented by default is setting a generic message at boot with the Debian Version on it. 

I've got a separate project with examples for setting the LCD including examples:
https://github.com/1000001101000/Python_buffalo_libmicon

**Reply by jhmc93** [reply]
is it easy to install?
ur bookworm install do i transfer initrd.buffalo.armel with the armel at the end or without?

**Reply by 1000001101000** [reply]
assuming the first half of question is about the python library.

There is a copy of the python library under /usr/local/bin if you install via my installers. You can also get the repo with git clone etc.

The process for loading the installer files is the same for every version, the files must always be renamed to what the device expects (initrd.buffalo and uImage.buffalo)

**Comment by jhmc93** [comment]
do u have it the lcd screen installer

**Reply by 1000001101000** [reply]
Click on the link I posted earlier

**Reply by jhmc93** [reply]
I've got problems, my 1st installer messed up because my internet died during install, ive reformatted the drive transferred the bookworm stuff over but it keeps going into emergency mode

**Reply by 1000001101000** [reply]
It's mostly a matter of repeating the steps that were successful earlier.

If you provide enough detail about the steps you are taking I might be able to spot where things are going wrong.

**Comment by jhmc93** [comment]
i start with pulling the bookworm files u gave me through the simple http server onto /boot
due to the files being name initrd.buffalo.armel i use mv 'initrd.buffalo' and with uImage i use 'mv uImage.buffalo.tsxl to uImage.buffalo' then i do 'sync' then 'exit' and then the java command to restart and it goes straight to emergency mode

**Comment by 1000001101000** [comment]
if the device is in emergency mode /boot will be in a ramdisk rather than on your drive, if you run `mount /dev/sda1 /boot` before doing the file transfer that might solve things.

**Reply by jhmc93** [reply]
i've tried that but it keeps unmounting on each reboot...

**Reply by 1000001101000** [reply]
it's not something that persists from boot to boot, but if the next restart it boots into the installer it also won't matter. 

Either you successfully did that once last time you got the installer working or you connected the drive to a pc or something and loaded it that way..... either way whatever you did last time should work this time.

**Reply by jhmc93** [reply]
well i've tried nothing works, i've tried even using tsupdater and install the debian over the buffalo stock but nothing is working, i tried a different hard drive to, each time it a red screen. it's a shame because i nearly had it first time, so i don't know  whats going on, i've tried my hardest. But no installer

**Reply by 1000001101000** [reply]
There are a limited number of things that could be going on. Get back into em mode and mount /boot again , when you get that far show the output of `df` and `mount` so we can confirm some details about your /boot. then once you copy the files over run `ls -lh /boot` to give us some details about your boot files.

**Comment by jhmc93** [comment]
![df](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/9238e307-dfd5-4e5f-818a-bd19449c3a22)
![ls lh](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/ca2c1adf-1d9d-4108-863b-9036e1033e59)
![mount](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/6ebd10d5-2d6c-491a-bfe8-7a77651e3ebd)
here

**Reply by 1000001101000** [reply]
what version are you trying to install right now?

**Reply by jhmc93** [reply]
the bookworm.zip u gave me

**Reply by 1000001101000** [reply]
actually, could you run `date` to confirm what the system clock says? Those don't look like the right boot files either way but November seems like an odd timestamp.

**Comment by jhmc93** [comment]
![Screenshot_3](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/5faca1d0-22f3-4458-ab40-d7b197f1da30)

**Reply by 1000001101000** [reply]
that filesize for the initrd.buffalo looks about right, that uImage.buffalo is way off though. 
Could you run:
`md5sum /boot/uImage.buffalo`
`md5sum /boot/initrd.buffalo`

**Reply by jhmc93** [reply]
![Screenshot_4](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/43d03b75-c9de-46dd-b5a9-97a6e19bc639)

**Reply by 1000001101000** [reply]
This is the files from the zip:
40af97d98fdde32b8af066176623260f  initrd.buffalo.armel
d3124fca7e4d7d964f64632c18f6a1f5  uImage.buffalo.tsxl

looks like you’ve got the wrong uImage somehow, once you load matching files you should have better luck

**Reply by jhmc93** [reply]
right this is before i reboot 
![Screenshot_5](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/a6011f14-4715-4a8f-9b0e-b641e54cab7a)

**Reply by jhmc93** [reply]
![Screenshot_6](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/852983cd-bf7d-4f82-baf9-82635523c434)
still emergency mode

**Reply by 1000001101000** [reply]
please provide more details between screen shots, what exactly have you done since it went into em mode again? I'll assume you mounted sda1 on /boot again and confirmed what's there has the expect md5sums

could you run `fdisk -l /dev/sda` ?
also `dumpe2fs /dev/sda1 | grep -i size` (which might not work).
also does `smartctl -a /dev/sda` return anything?
also, are there any other drives connected?

if we're sure we have the right files that's one piece. the filenames look correct, if there is some subtle typo I don't see it. 
your `mount` output earlier confirmed it is formatted ext3. 
your df output seemed to show it having a plausible size.

by best guess is an issue with the partition table, though I don't know how that would have happened between when it was working and now, possibly something with your attempted stock firmware install I suppose. Let me see the output of the above commands, if nothing useful comes of that we can wipe the drive and set up /boot all from EM mode.

**Reply by jhmc93** [reply]
Sorry, I have mounted the /boot replaced the uImage and initrd in the /boot and then i ran the md5sums after I replaced them, hence this screenshot:
> right this is before i reboot ![Screenshot_5](https://user-images.githubusercontent.com/50001717/244543222-a6011f14-4715-4a8f-9b0e-b641e54cab7a.png)

after rebooting with this command: "java -jar acp_commander.jar -t 192.168.1.2 -pw password -c "reboot"" 
It went straight back into emergency mode,
so I started the telnet service again and mounted the /boot, I then ran the md5sums command again hence this screenshot:

> ![Screenshot_6](https://user-images.githubusercontent.com/50001717/244545281-852983cd-bf7d-4f82-baf9-82635523c434.png) still emergency mode

the output of the commands below:

- fdisk -l /dev/sda

- dumpe2fs /dev/sda1 | grep -i size
 
-  smartctl -a /dev/sda

are in the below screenshot:
![output](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/dbf6577c-8c97-4505-a893-0b7848c222d2)

also there is only 1 320gb drive in it at the moment with the one partition for the /boot

**Reply by 1000001101000** [reply]
the inode size needs to be 128, that might be what the problem is. how did you last format that partition? Like what tool in what OS?

you might be able to just reformat sda1 and reload the files unless there are other issues lurking.

Might be as simple as:
`umount /dev/sda1`
`mkfs.ext3 -I 128 /dev/sda1` (say yes to prompts)
then remount and copy the boot files in again. before rebooting run the dumpe2fs command again to verify the inode size

**Reply by jhmc93** [reply]
I really appreciate your time helping me, I'll have a go at this tomorrow, right now though I'm off to bed as I live in the UK and it's 2:44am here, so I'll catch u tomorrow and let you know how its getting on, also I used the drive with fdisk to partition and format

**Reply by jhmc93** [reply]
dumpe2fs output:
![Screenshot_7](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/33371f7a-468f-4d12-b2e5-7ba1c19a1b05)

![Screenshot_8](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/60f6dfd7-737f-47ea-8e5d-144b85eb100e)

**Reply by 1000001101000** [reply]
that looks more sensible

**Reply by jhmc93** [reply]
EM mode again

**Reply by 1000001101000** [reply]
did you load the files before rebooting?

**Reply by jhmc93** [reply]
i didnt mv initrd.buffalo.armel to initrd.buffalo and uImage.buffalo.tsxl to uImage.buffalo,
even though i've done the mv it still looks like it gonna go in EM mode

**Reply by jhmc93** [reply]
PORT 22  ssh is OPEN :D

**Reply by jhmc93** [reply]
![Screenshot_9](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/dccdc583-929d-4188-8ddf-2a7f17c37cc2)
do i need to remove anything?

**Reply by jhmc93** [reply]
nvm, it's installed but i need your help, have i done this right i went into the github u sent me and cd into the folder, then ran the below .py and hit enter but nothing happened
 

![Screenshot_10] (https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/a619450c-987b-492d-8827-e528b8ff9ad6)

**Reply by 1000001101000** [reply]
that is a script could use a better name. It's an example of a script that watches for the power button being held down and then shuts down the device if it is held for > 2sec. It's not expected to do anything except in that scenario.

`ts3000_examples.py` will cycle through most of the available micon operations and can be used as example code for building whatever you desire.

`button_test.py` runs in a loop and displays what buttons are pressed whenever that changes. It can be a little confusing since the buttons are only referenced by number and buttons not present on the device always show as pressed. with a few tweaks it could be used for something useful.

**Reply by jhmc93** [reply]
so can u go through the install with me? to make it easier for me as i'm new to python scripts

**Reply by 1000001101000** [reply]
I don't really understand what install means in this context. you managed to run button_daemon.py, the process is the same for any of the other scripts.

clone the repo if you haven't already then run `./ts3000_examples.py`

**Reply by jhmc93** [reply]
button_daemon.py ended up running into erros after 5 minutes

**Reply by 1000001101000** [reply]
I'm not surprised. I haven't implemented anything to ensure different scripts don't try to talk to the micon at the same time, probably that button script and the fan_daemon tried at the same time and hit an issue. 

I only added the fan_daemon because some of the rackmount models sound like a jet taking off if you don't turn down the fans after a boot. 

The whole thing could use someone more familiar with that type of programming to set up something more robust though as-is it's usually "good enough" to make these devices usable which is mostly what I targeted when doing that bit of work.

**Reply by jhmc93** [reply]
So do a fresh restart and try button_daemon.py again

**Comment by 1000001101000** [comment]
why are you trying to run button_daemon ?

**Reply by jhmc93** [reply]
that's how I though u implemented it into the system so it gets stuff working :/ sorry i sound thick

**Comment by 1000001101000** [comment]
The only things I've implemented in the sense that they are set up by the installer:

- micon_startup.py: sends signal to disable boot watchdog once system loads, sets initial LCD message
- micon_shutdown.py: sends signals needed for device to shutdown/restart properly.
- micon_fan_daemon.py: attempts to read temperature and make fan speed adjustments.

everything else in the repo are examples that can be used as a starting point for anything you desire. Honestly, even the ones used by the installer are glorified example scripts, bare minimum to make the device run properly.

**Reply by jhmc93** [reply]
So what should I run? This: ./ts3000_examples.py

**Reply by 1000001101000** [reply]
Yes.
<img width="375" alt="image" src="https://github.com/1000001101000/Debian_on_Buffalo/assets/38921757/3125555b-8e18-42e9-864b-337d5051fbc2">

**Reply by jhmc93** [reply]
What does the 3000 stand for btw?

**Reply by 1000001101000** [reply]
It's an example written to run on the TS3000 model devices which is the closest to the TS-XL in terms of micon functionality of the devices I wrote examples for

**Reply by jhmc93** [reply]
![image](https://github.com/1000001101000/Debian_on_Buffalo/assets/50001717/55c9350c-530b-41ef-b62a-7830545f6e8d)
what does this mean

**Reply by 1000001101000** [reply]
sounds like best practices for signing repository stuff have shifted since I set mine up, probably something I should look into someday

**Reply by jhmc93** [reply]
ok thank you, I'm just downloading a Desktop environment onto it now, my front display works now so thanks for your help on this

**Reply by jhmc93** [reply]
hello again, quick question, the ./ts3000_examples.py command doesn't work at startup, is there a way to fic this

**Reply by 1000001101000** [reply]
can you elaborate on what you mean?

What did you do to try to make it run at startup? How did you determine that didn't work?

**Reply by jhmc93** [reply]
So I ran the python script as mentioned above and the display button on the terrastation it scrolls through the options for example, ip address, Debian version, time and date, but after a reboot of the terrastation  it stays on the Debian arm 12.0 screen if I press the display button nothing happens

**Reply by 1000001101000** [reply]
The code that sets the LCD at boot is triggered by a systemd service that launches it during the boot process. 

You can find the script itself and the .service files that I used in the repo:
https://github.com/1000001101000/Debian_on_Buffalo/tree/master/Tools/micon_scripts

You might be able to accomplish what you want by customizing that script/service or making something similar based on your needs.

**Reply by jhmc93** [reply]
So how would I get it to run the Python script at boot?

**Reply by 1000001101000** [reply]
Take a look at the systemd service I use to start mine at boot (see link in my previous post). It may more sense if you lookup info about using systemd service files for startup things like this.

**Reply by jhmc93** [reply]
would a cronjob work?

**Reply by 1000001101000** [reply]
I would think so

**Reply by jhmc93** [reply]
Nvm I used a systemd script, all working, can u use the second Ethernet on the terrastation ts-xl with the Debian install

**Reply by 1000001101000** [reply]
yes, you can do whatever you like with them

**Reply by jhmc93** [reply]
Is it possible to forward traffic on one port

**Reply by 1000001101000** [reply]
It is possible to do anything that is supported by networking under linux, there are far better resources out there to learn about the details than me.

**Comment by Nick7677** [comment]
Hello, i have a TS-XL/R5 and i'm stuck after loading the initrd.buffalo and uImage.buffalo  files. 
The system after the reboot, stays in "terastation arm Debian Installer" message in the LCD display.  
I can't do anything: no ssh , no acp-command , no nas navigator access. 
Specifically, when i try an ssh installer@ip.address  it says the port 22 is closed. 
Can you help me or give me a clue? thanks in advance

**Reply by 1000001101000** [reply]
I’d scan the network for devices with port 22 open. It could have a different IP than you expect

**Reply by Nick7677** [reply]
the ip address is correct: i can see it through the router cause i've had the same thinking the first time i tried the installation.
actually, the ssh connection says "rejected". 
Could it be a newer firmware ( 1.76 )of the NAS that causes the issue?  
thanks for you support

**Reply by 1000001101000** [reply]
The firmware version won’t affect what the installer does. The device does need to download the ssh components from the debian mirror before it can proceed to that stage. If it doesn’t have an internet connection that could explain it.

a momentary failure to reach the mirror could even do it, powering off and trying again might make it work

**Reply by Nick7677** [reply]
That's why! i was working offline... now it's working. Thanks a lot for your great job!

**Reply by 1000001101000** [reply]
Cheers!
